### ISSUE TITLE
RSA decryption primitive testing: Unclear test approach

### ISSUE THREAD

**Issue by smuellerDD** [usnistgov/ACVP#750]
According to SPEC, for the RSA decryption primitive the IUT is supposed to generate a new key and perform the decryption operation.

However, allowing the IUT to generate the key, the base requirement that the cipher text is < p -1 is not a fixed requirement.

I have implemented that with OpenSSL and every time I perform a new test, I get different numbers of failing test results, i.e. I see varying numbers of testPassed:false results. Each testPassed:false is due to the OpenSSL failure "Decryption failed error:04065084:rsa routines:rsa_ossl_private_decrypt:data too large for modulus" - i.e. the base requirement above is a moving target for a given cipher text but a varying key.

In addition, the SPEC does not provide descriptions about the different JSON keywords defined for the decryption primitive, e.g. totalFailingCases.

Also, the test vector with a "resultsArray" structure is a major structural departure from all other vectors.

**Comment by celic** [comment]
I don't recall liking how this test was set up. The reason for the `resultsArray` is because on our end we needed to validate all the test cases at once to ensure that a certain number of failures occurred. At the time we didn't have a way of verifying multiple individual test cases as a unit, but now results have the test group structure. This could probably be updated, but if we were to work on this, I'd rather solve the problem that the tests don't provide much of an assurance in the first place. 

Their purpose is to see that an implementation throws that failure you mentioned. How they do it is a moving target, yes. We dump the work on the client to keep generating keys and match them up until they satisfy the expected number of failures. We had a similar problem for the signature primitive tests but the work was on the server side instead. We avoided that by generating "fake" keys, setting the upper bits to ensure a failure. You have mentioned issues with that approach as well. 

These two tests use the same principle and it is a bit flawed. We could spend some time iterating on their design while still getting the assurances we want.

**Comment by smuellerDD** [comment]
Am Mittwoch, 11. Dezember 2019, 16:41:10 CET schrieb Chris Celi:

Hi Chris,

> I don't recall liking how this test was set up. The reason for the
> `resultsArray` is because on our end we needed to validate all the test
> cases at once to ensure that a certain number of failures occurred. At the
> time we didn't have a way of verifying multiple individual test cases as a
> unit, but now results have the test group structure. This could probably be
> updated, but if we were to work on this, I'd rather solve the problem that
> the tests don't provide much of an assurance in the first place.

Fine with me, this structural difference was FYI only.
> 
> Their purpose is to see that an implementation throws that failure you
> mentioned. How they do it is a moving target, yes. We dump the work on the
> client to keep generating keys and match them up until they satisfy the
> expected number of failures.

Ahh, you say that the client needs to retry as often to get the requested 
number of failures?!

> We had a similar problem for the signature
> primitive tests but the work was on the server side instead. We avoided
> that by generating "fake" keys, setting the upper bits to ensure a failure.
> You have mentioned issues with that approach as well.
> 
> These two tests use the same principle and it is a bit flawed. We could
> spend some time iterating on their design while still getting the
> assurances we want.

This is for the future. As of now I am more interested in getting the testing 
complete.


Ciao
Stephan

**Comment by smuellerDD** [comment]
After changing the code to try to generate the exact number of failures, I am unable to generate the requested number of failures.

10 failures are requested, but I always get more, even when I generate an RSA key up to 10 times for each cipher text.

**Comment by smuellerDD** [comment]
Even with more failures than requested, I generated a JSON response that looks compliant to the SPEC, the ACVP server complains about the format.

JSON test result:

```
[
  {
    "acvVersion":"1.0"
  },
  {
    "vsId":90545,
    "testGroups":[
      {
        "tgId":1,
        "tests":[
          {
            "e":"daaa316f",
            "n":"9f451ca07fc405c7e15f6e174d19b99736776684ec709c9ddc9d5b8b9706704fb4dc2cbd4ccf92f02f3cf8f8e89f37c766b789a5c57ee2a3e1485be9967fd2370a5d0be50e0788c5c3fe9b9f8199ae330b901f911b5268ddfd676a8ee85dc53fde9b506222ce470b27ffe9b7c94e78b0f09617bbf2b55e732c936f39534bc4885e5f73aaf60f26e2af958487c1aeb6985708ce41d8327cdcb681f1ecbb847a6a0b2915f17366542135fe14fd7c943ac290f062028fc0f0cd7c48926aa29f7777bbcea5fbe940f5dca6873ac7920e64743d0bcc934e0d319c0b6764abc2b5ef5d42809ad068ac1cbf5303faacfe945e0458f7a01fe0797af908f1c149242ddfd1",
            "tcId":1,
            "plainText":"9c44bd30f9b0bc5c9aefc9ae7f1210f689c8e11b8238e266d8f33a8bde1373b3731cfa8633e9485e64d15cdd7d5b5d453356b3fdbcc76da7f4caebcbc6b9af4884e65100380e5aee8660b649798945150ff1888ee6b196498c5995c64dba2cfd820424203e37b1f7e4aeb7833d448c9656513d526e5063485b521a1de9930f771c78a7a00cee41954f932229009c9d1ba059e8d254641a006a317f27e62fbc5837f927f4b6263c5177b2c65f967ee66580c42d21d9b4a7857a4ecf00dd6e54a051c373a4502641b2b5809917c57877f739b697976fe9955e5417424db876a81ec640be38e911efd62a38a60247b1843dc574e93454497c4b955c362cf7f981a7"
          },
```

The server answers:

```
{
  "acvVersion" : "1.0",
  "error" : "Incorrectly formatted JSON (194:4): Unknown or disallowed submission property: e"
}
```

**Comment by Kritner** [comment]
@smuellerDD the response is expected in a format more similar to an MCT response like:

```
{
  "vsId": 0,
  "algorithm": "RSA",
  "mode": "decryptionPrimitive",
  "revision": "1.0",
  "testGroups": [
    {
      "tgId": 1,
      "tests": [
        {
          "tcId": 1,
          "resultsArray": [
            {
              "e": "60BDBEF656869D",
              "n": "8FA73CF9CAD37456B64B3B3DF75C3D3BF254A62C82F445682D0BC34FC998F893039C964E3F3B2F0BD70AA39FB693AD5E1C29398BCE7D43A6F57C34FADF4C6159EBF2D1A4BB5A652BDF74A9C69A3AE46105A29B2AF2E385D54152A8A4660F8081D03DDA9AF5B301B8542B6E535285F89D219A095FCD3296C58DC758BC12B9564EC8FC4B92D805FC0F01695D89A9129C9A0EBB5EBC5D487D1CD0B3A0F2C30321B1B41766EF1F0659805667A84B4F66792DB91BBF346B0A652FEB6B9932855377AAB4ACF0224056B6CEF0CAC7C378698869E526453AADD65EA43AA746D5D5494A1E2A20B4D7D05F53FF566C0BC9AFA0D731416E7BD071A2CA6984C08294560D3BFB",
              "testPassed": false
            },
```

This syntax is a bit clunky to be sure, but as @celic stated, hopefully we'll be able to revisit it at some point.

**Comment by smuellerDD** [comment]
Ok, I updated my output generator accordingly. Yet, I think I cannot test this as the initial issue is yet to be clarified.

**Comment by smuellerDD** [comment]
May I ping on this one to get an idea about the initial issue?
Thank you.

**Comment by dmchodnicki** [comment]
ACVP Technical Team has decided to not do much on Decryption Primitive tests. Overall, the tests are not that useful and are not expected to be used with any frequency.  No action is currently planned and this item is being Closed.

**Comment by nidhidamodaran** [comment]
> ACVP Technical Team has decided to not do much on Decryption Primitive tests. Overall, the tests are not that useful and are not expected to be used with any frequency. No action is currently planned and this item is being Closed.

@dmchodnicki Does that mean these tests are no longer relevant ?

**Comment by celic** [comment]
The RSADP tests have a very specific scope. They are only applicable to modules that seek NPIVP validations. No other module should be performing these tests. In reality this is about 2-3 modules a year that should be running these tests. So the tests are still relevant, just only in specific cases.

**Comment by kaleb-himes** [comment]
@celic,

We have multiple module versions and many OEs' on cert [3389](https://csrc.nist.gov/projects/cryptographic-module-validation-program/certificate/3389). We are adding more OEs' over time. We run this test 2-3 times a month, it is not a rare test (in our case anyway). Right now one of the feasible ways to implement the test with the hard count for success and failures is to use:

```
            "totalTestCases": 30,
            "totalFailingCases": 10,
```

1) Track the number of failures during processing and once the `totalFailingCases == numberOfFailures`, any future tests that fail, run them over and over and over .... and over and over until they pass. This is the most common case being executed by our test harness.

2) If we reach the end of the vector set and `totalFailingCases > numberOfFailures`, re-run the file again (up to X number of retries (I have our ceiling at 30 retries but I've never needed more than 2 when re-running the entire file). This is very rare and I have only seen it happen when processing DEMO vectors, never on PRODUCTION vectors.

It is not uncommon to encounter cipherTexts that take 250 (+/- 50) attempts before we get a successful decrypt once we have enough failures. However, I just ran into a vector set that has a cipherText that failed to decrypt after 500,000+ attempts. I finally had to write our FIPS lab and tell them we got a bad set from the ACVP server and request that we receive a new PRODUCTION set. In the 50+ DEMO and PRODUCTION sets I've run this year, the above has never happened before. This one case does seem to indicate that there needs to be an improvement to this test approach. Perhaps the ACVP server should at least confirm the cipherText can be decrypted before sending off the requests to be processed by the IUT?
